ansible/roles/linux-ansible-init/files/ansible-init.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/lib/ansible-init/bin/python
+
+import json
+import logging
+import os
+import pathlib
+import subprocess
+
+import requests
+
+
+logging.basicConfig(level = logging.INFO, format = "[%(levelname)s] %(message)s")
+
+
+logger = logging.getLogger(__name__)
+
+
+def assemble_list(data, prefix):
+    """
+    Assembles a list of items based on keys in data with the format "<prefix>_<idx>_<key>".
+    """
+    list_items = {}
+    for key, value in data.items():
+        if not key.startswith(prefix):
+            continue
+        idx, item_key = key.removeprefix(prefix).split("_", maxsplit = 1)
+        list_items.setdefault(idx, {})[item_key] = value
+    return [list_items[k] for k in sorted(list_items.keys())]
+
+
+def ansible_exec(cmd, *args, **kwargs):
+    """
+    Execute an Ansible command with the appropriate environment.
+    """
+    environ = os.environ.copy()
+    environ["ANSIBLE_CONFIG"] = "/etc/ansible-init/ansible.cfg"
+    cmd = f"/usr/lib/ansible-init/bin/ansible-{cmd}"
+    subprocess.run([cmd, *args], env = environ, check = True, **kwargs)
+
+
+logger.info("fetching instance metadata")
+METADATA_URL = "http://169.254.169.254/openstack/latest/meta_data.json"
+response = requests.get(METADATA_URL)
+response.raise_for_status()
+user_metadata = response.json().get("meta", {})
+
+
+logger.info("extracting collections and playbooks from metadata")
+collections = assemble_list(user_metadata, "ansible_init_coll_")
+playbooks = assemble_list(user_metadata, "ansible_init_pb_")
+logger.info(f"  found {len(collections)} collections")
+logger.info(f"  found {len(playbooks)} playbooks")
+
+
+logger.info("installing collections")
+ansible_exec(
+    "galaxy",
+    "collection",
+    "install",
+    "--force",
+    "--requirements-file",
+    "/dev/stdin",
+    input = json.dumps({ "collections": collections }).encode()
+)
+
+
+logger.info("executing remote playbooks for stage - pre")
+for playbook in playbooks:
```

Comment:
Technically don't need to loop here if you splat a sorted list of playbook paths into a single `ansible-playbook` command.

Reply:
I quite like making sure that we have control over the ordering (in this case they will execute in order of the index in the metadata item)